napari/_qt/qt_plugin_list.py

```diff
@@ -0,0 +1,199 @@
+            NapariPluginManager instance
+        parent : QWidget, optional
+            Optional parent widget, by default None
+        initial_hook : str, optional
+            If provided the QComboBox at the top of the dialog will be set to
+            this hook, by default None
+        firstresult_only : bool, optional
+            If True, only hook specifications that declare the "firstresult"
+            option will be included.  (these are hooks for which only the first
+            non None result is returned).  by default True (because it makes
+            less sense to sort hooks where we just collect all results anyway)
+            https://pluggy.readthedocs.io/en/latest/#first-result-only
+        """
+        plugin_manager = plugin_manager or napari_plugin_manager
+        super().__init__(parent)
+        self.plugin_manager = plugin_manager
+        self.layout = QVBoxLayout()
+        self.setLayout(self.layout)
+        self.hookComboBox = QComboBox()
+        self.hookComboBox.addItem(self.NULL_OPTION)
+
+        # populate the combo box with all of the hooks known by the plugin mngr
+        hooks = []
+        for name, hook_caller in vars(plugin_manager.hook).items():
+            if firstresult_only:
+                # if the firstresult_only option is set
+                # we only want to include hook_specifications that declare the
+                # "firstresult" option as True.
+                if not hook_caller.spec.opts.get('firstresult', False):
+                    continue
+            hooks.append(name)
+        self.hookComboBox.addItems(hooks)
+
+        self.hookComboBox.activated[str].connect(self.change_hook)
+        self.hookList = QtHookImplListWidget()
+
+        title = QLabel('Plugin Sorter')
+        title.setStyleSheet(
+            "QLabel { background-color : rgb(38, 41, 47); "
+            "color : #ddd; font-size: 18pt; }"
+        )
+        self.layout.addWidget(title)
+
+        descript = QLabel(
+            'select a hook to rearrange, then drag and \n'
+            'drop plugins into the desired call order'
+        )
+        descript.setStyleSheet(
+            "QLabel { background-color : rgb(38, 41, 47); "
+            "color : #bbb; font-size: 12pt; }"
+        )
+        self.layout.addWidget(descript)
+
+        self.layout.addWidget(self.hookComboBox)
+        self.layout.addWidget(self.hookList)
+        if initial_hook is not None:
+            self.hookComboBox.setCurrentText(initial_hook)
+            self.change_hook(initial_hook)
+
+    def change_hook(self, hook: str) -> None:
+        """Change the hook specification shown in the list.
+
+        Parameters
+        ----------
+        hook : str
+            Name of the new hook specification to show.
+        """
+        if hook == self.NULL_OPTION:
+            self.hookList.set_hook(None)
+        else:
+            self.hookList.set_hook(getattr(self.plugin_manager.hook, hook))
```

Comment:
This line is confusing to me... Does `self.plugin_manager.hook` contain attributes for every possible hook name?

Reply:
This might make more sense to me:

```python
getattr(self.plugin_manager.hooks, hookname)
```
? Am I parsing this right?